### PR TITLE
Add UBI image support

### LINKED ISSUE #1871: Conjur builds a UBI-based image
Conjur should build a [UBI-based](https://developers.redhat.com/products/rhel/ubi) image that can be used when running Conjur in OpenShift and can ultimately be published to the RH container registry.

This is related to cyberark/conjur-oss-helm-chart#60.

### PR DESCRIPTION
**Pull request by guygiat**
### What does this PR do?
- New Dockefile.ubi with relevant packages and permissions

### What ticket does this PR close?
Resolves #1871 

### Checklists
- [x] Add LICENSE file to the main repo project
- [x] Add `COPY LICENSE /licenses` to the Dockerfile.ubi
- [x] Change relevant description and summary on the Dockerfile.ubi labels section 
- [x] Adding a build step for the new image to the Jenkinsfile
- [x] Adding a separate commit to bump the project Ruby version to 2.5.5

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [x] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

### CODE REVIEW

**Comment by izgeri** [comment]
@guygiat can you add commands to build this image in the [build script](https://github.com/cyberark/conjur/blob/master/build.sh)?

**Comment by izgeri** [comment]
Should this be marked as `WIP` in the PR title? It looks like you have a checklist for some additional items that you want to include in this PR

**Comment by guygiat** [comment]
@izgeri the checklist is also for your team, I think they will need to review those stuff, for example I don't know what to put in the LICENSE file and the labels part

File: Dockerfile.ubi
Comment (sgnn7) [inline Dockerfile.ubi:27]:
```suggestion
    INSTALL_PKGS="ruby ruby-devel rubygem-rake redhat-rpm-config libxslt-devel libxml2-devel gcc gcc-c++ glibc-devel make git openldap-clients libpq libpq-devel tzdata libzip-devel" && \
    yum install -y --setopt=tsflags=nodocs gcc \
                                           ... \
                                           rubygem-rake \
                                           tzdata && \
```

Comment (sgnn7) [reply Dockerfile.ubi:27]: Or split the package names the same way if you really need `rpm -V`

File: Dockerfile.ubi
Comment (sgnn7) [inline Dockerfile.ubi:42]: I still don't know why `docker/` can't be used though

File: Dockerfile.ubi
Comment (sgnn7) [inline Dockerfile.ubi:44]: This version number should be a variable up top

File: Dockerfile.ubi
Comment (sgnn7) [inline Dockerfile.ubi:25]:
```suggestion
RUN yum -y module enable "ruby:$RUBY_VERSION" && \
```

File: Dockerfile.ubi
Comment (sgnn7) [inline Dockerfile.ubi:52]:
```suggestion
RUN mkdir -p "$TMP_DIR" \
             "$LOG_DIR" \
             "$SSL_CERT_DIRECTORY/ca" \
             "$SSL_CERT_DIRECTORY/cert" \
```

**Review by sgnn7 — CHANGES_REQUESTED** [review]
@guygiat Mostly nit comments. It would be good that someone from the core team also takes a look too.

File: Dockerfile.ubi
Comment (izgeri) [inline Dockerfile.ubi:23]:
```suggestion
      summary="Conjur UBI-based image" \
      description="Conjur provides secrets management and machine identity for modern infrastructure. "
```

**Review by izgeri — CHANGES_REQUESTED** [review]
@guygiat here are some items I'd like to see done before we merge this:
- [x] The license file for this project is available at https://github.com/cyberark/conjur/blob/master/LICENSE.md - can you please add a step to the Dockerfile that does:
  ```
  ADD LICENSE.md /licenses
  ```
- [ ] Add commands to build this image in the [build script](https://github.com/cyberark/conjur/blob/master/build.sh)

Is the Ruby version change the only one needed to satisfy "bump the project Ruby version to 2.5.5"?

File: Dockerfile.ubi
Comment (guygiat) [reply Dockerfile.ubi:42]: actually we don't need this anymore, I forgot to remove this line.

File: Dockerfile.ubi
Comment (guygiat) [reply Dockerfile.ubi:23]: for the LICENSE part, I'm not sure the md format can work with the Redhat registry.
In our other Redhat repos we use a LICENSE which isn't with md extension and it looks a little bit different.
for example: https://github.com/cyberark/secretless-broker/blob/master/LICENSE

Comment (izgeri) [reply Dockerfile.ubi:23]: other projects use Apache 2.0, but this project uses LGPL. I think we should move forward with the license as-is for now, and we can modify if needed when we start publishing to RH and find out we have issues. I don't see anything in their [docs](https://redhat-connect.gitbook.io/partner-guide-for-red-hat-openshift-and-container/certify-your-application/creating-a-container-application-project) that indicates that we should have an issue with a MD license file, fwiw.

Comment (izgeri) [reply Dockerfile.ubi:23]: [this page](https://connect.redhat.com/zones/containers/container-certification-policy-guide) just says:
> How? Create a directory named /licenses and include all relevant licensing and/or terms and conditions as text file(s) in that directory

Comment (guygiat) [reply Dockerfile.ubi:23]: not a ruby person so not sure this is the only change needed, but it solves the issue when building the image

File: Dockerfile.ubi
Comment (izgeri) [inline Dockerfile.ubi:85]: can you verify that when the image is built, it respects the repo [.dockerignore](https://github.com/cyberark/conjur/blob/master/.dockerignore) and doesn't include the `design` and `docs` folders or any of the `.sh` files?

File: build.sh
Comment (izgeri) [inline build.sh:73]: is this image also going to need the `flatten` command that we use for the Conjur image above?

File: Gemfile
Comment (izgeri) [inline Gemfile:12]: @cyberark/conjur-core-team is this single change sufficient to bump the base Ruby version in this project / in its docker images?

**Review by izgeri — CHANGES_REQUESTED** [review]
a few additional comments, and one ask for more info on the Ruby change from the Conjur Core team

File: build.sh
Comment (guygiat) [reply build.sh:73]: I'm not sure why you are doing this, the image is working perfectly for me without loosing any env variables, worth to ask @cyberark/conjur-core-team as well.

File: Dockerfile.ubi
Comment (guygiat) [reply Dockerfile.ubi:85]: yeah it respects it, but it seems that we copy a lot of other unnecessary files to the image (as well for the original conjur image)

Comment (izgeri) [reply Dockerfile.ubi:85]: if you have other suggestions for what to put in the .dockerignore feel free to add those changes (in this PR or in a follow-on PR)!

File: build.sh
Comment (izgeri) [reply build.sh:73]: based on the [version](https://hub.docker.com/r/guygiat/conjur-ubi-ocp/tags) you pushed to dockerhub (207 MB) vs the size of the `cyberark/conjur` [image on dockerhub](https://hub.docker.com/r/cyberark/conjur/tags) (257 MB), I don't think you need the `flatten` here :)

**Comment by sgnn7** [comment]
Drive-by note here since I missed a day:
- History of this PR will need cleanup. We also have a master branch merge onto this branch which will need to be removed.
- I checked and this image has 12 layers when pulled which makes it very untidy and it's size is 571MB if you use [`dive`](https://github.com/wagoodman/dive) to examine it (pic attached). It's not mandatory to have flattening added to it but it would be real nice.

Screen Shot 2020-10-09 at 07 59 05

**Comment by izgeri** [comment]
@guygiat we'll be out until Tuesday for the US holiday on Monday, but if you get this fixed up by then I think it'll be in a good place to get merged.

**Comment by guygiat** [comment]
@sgnn7 After I flatten my image it total size is just **556 MB**, do you think we still need the flatten?
The original `cyberark/conjur` image is **671 MB** after flatten.

**Comment by sgnn7** [comment]
> @sgnn7 After I flatten my image it total size is just 556 MB, do you think we still need the flatten? The original cyberark/conjur image is 671 MB after flatten.

I built the images both ways and you're right, we won't save any space on flattening it would just be a bit neater/tidier so we can probably skip it for now. 

However, I think we should probably use `FROM registry.access.redhat.com/ubi8/ubi` as it doesn't require authentication: https://access.redhat.com/RegistryAuthentication

**Comment by guygiat** [comment]
@sgnn7 I agree to go with `registry.access.redhat.com` but from the link you sent it is going to be decommissioned this year:

> Over the coming year, Red Hat will standardize on registry.redhat.io and registry.connect.redhat.com as the primary container registries for Red Hat and certified partners, and will eventually decommission registry.access.redhat.com. While customers can continue to use registry.access.redhat.com until it is decommissioned, it is recommended that you prepare to use registry.redhat.io. 

Do you want to move on with `registry.access.redhat.com` for now?

**Comment by sgnn7** [comment]
@guygiat hmm neither option is good tbh - maybe this is something @cyberark/conjur-infra-team can help figure out?

File: Dockerfile.ubi
Comment (sgnn7) [inline Dockerfile.ubi:42]: Please sort the packages aplhabetically

File: Dockerfile.ubi
Comment (sgnn7) [inline Dockerfile.ubi:89]: This should be much earlier in the file. Also, you should use `COPY` instead of `ADD` for a local file

**Review by sgnn7 — CHANGES_REQUESTED** [review]
@guygiat Just a couple of comments but it would be good that this is looked over by someone from the @cyberark/conjur-core-team too.

File: Dockerfile.ubi
Comment (guygiat) [reply Dockerfile.ubi:89]: Changed to COPY, all of our other repos put the license file at the end of the Dockerfile that is why I put it at the same place :)

File: Gemfile
Comment (sgnn7) [reply Gemfile:12]: @cyberark/conjur-core-team ping

File: Dockerfile.ubi
Comment (sgnn7) [reply Dockerfile.ubi:89]: @guygiat I still see `ADD` here. License is perfectly fine to add early as it can then be cached rather than re-copied each time as a new layer anything in the repo changes.

File: Dockerfile.ubi
Comment (sgnn7) [reply Dockerfile.ubi:42]: This list is still not in alphabetic order

**Comment by sgnn7** [comment]
A note here that whoever merges this change in that they should probably squash-merge to clean up the git history.

@guygiat The branch will also need a rebase as it's out-of-date w/ master branch.

**Review by sgnn7 — APPROVED** [review]
@guygiat LGTM!

**Comment by izgeri** [comment]
Closing this since it's moved to #1895

**Pull request closed without merging**
